File: python/leetcode_grind/day_456.py
# https://leetcode.com/problems/palindromic-substrings/description/?envType=daily-question&envId=2024-02-10
class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        ans = 0

        for i in range(n):
            dp[i][i] = True
            ans += 1

        for L in range(2, n+1):
            for left in range(n-L+1):
                right = left + L - 1
                
                if s[left] == s[right] and (L == 2 or dp[left+1][right-1]):
                    dp[left][right] = True
                    ans+=1

        return ans

File: src/leetcode_grind/day_456.rs
// https://leetcode.com/problems/palindromic-substrings/description
pub fn count_substrings(s: String) -> i32 {
    let s = s.chars().collect::<Vec<_>>();
    let n = s.len();
    let mut dp = vec![vec![false; n]; n];
    let mut ans = s.len() as i32;
    for i in (0..n).rev() {
        for j in i + 1..n {
            if s[i] == s[j] && (j - i <= 2 || dp[i + 1][j - 1]) {
                dp[i][j] = true;
            }
            ans += if dp[i][j] { 1 } else { 0 };
        }
    }
    ans
}
